app/components-react/windows/settings/Hotkey.tsx

```diff
@@ -0,0 +1,216 @@
+import { IHotkey, IBinding } from 'services/hotkeys';
+import { byOS, OS } from 'util/operating-systems';
+import { Inject } from 'services';
+import { Services } from '../../service-provider';
+import { MarkersService } from 'app-services';
+import { TextInput } from 'components-react/shared/inputs';
+import cx from 'classnames';
+import styles from './Hotkey.m.less';
+import HotkeyBinding from 'components-react/shared/HotkeyBinding';
+
+/**
+ * Represents a binding that has a unique key for CSS animations
+ */
+interface IKeyedBinding {
+  binding: IBinding;
+  key: string;
+}
+
+class HotkeyProps {
+  hotkey: IHotkey = {} as IHotkey;
+}
+
+// This is kind of weird, but the key attribute allows
+// us to uniquely identify that binding in the DOM,
+// which allows CSS animations to work properly.
+function createBindingWithKey(binding: IBinding): IKeyedBinding {
+  return {
+    binding,
+    key: Math.random().toString(36).substring(2, 15),
+  };
+}
+
+function getBlankBinding() {
+  return {
+    key: '',
+    modifiers: {
+      alt: false,
+      ctrl: false,
+      shift: false,
+      meta: false,
+    },
+  };
+}
+
+function addEmptyBindingAt(index: number) {
+  return (bindings: IKeyedBinding[]) => {
+    const newBindings = [...bindings];
+    newBindings.splice(index + 1, 0, createBindingWithKey(getBlankBinding()));
+    return newBindings;
+  };
+}
+
+function removeBindingAt(index: number) {
+  return (bindings: IKeyedBinding[]) => {
+    // If this is the last binding, replace it with an
+    // empty binding instead.
+    if (bindings.length === 1) {
+      bindings[0].binding = getBlankBinding();
+    } else {
+      bindings.splice(index, 1);
+    }
+
+    return bindings;
+  };
+}
+
+function setBindingAtIndex(index: number, binding: IBinding) {
+  return (bindings: IKeyedBinding[]) => {
+    const newBindings = [...bindings];
+    newBindings.splice(index, 1, createBindingWithKey(binding));
+    return newBindings;
+  };
+}
+
+/**
+ * Turns a binding into a string representation
+ */
+function getBindingString(binding: IBinding) {
+  const keys: string[] = [];
+
+  if (binding.modifiers.alt) keys.push(byOS({ [OS.Windows]: 'Alt', [OS.Mac]: 'Opt' }));
+  if (binding.modifiers.ctrl) keys.push('Ctrl');
+  if (binding.modifiers.shift) keys.push('Shift');
+  if (binding.modifiers.meta) keys.push(byOS({ [OS.Windows]: 'Win', [OS.Mac]: 'Cmd' }));
+
+  let key = binding.key;
+
+  const matchDigit = binding.key.match(/^Digit([0-9])$/);
+  if (matchDigit) key = matchDigit[1];
+
+  const matchKey = binding.key.match(/^Key([A-Z])$/);
+  if (matchKey) key = matchKey[1];
+
+  if (key === 'MiddleMouseButton') key = 'Mouse 3';
+  if (key === 'X1MouseButton') key = 'Mouse 4';
+  if (key === 'X2MouseButton') key = 'Mouse 5';
+
+  keys.push(key);
+
+  return keys.join('+');
+}
+
+export default function Hotkey(props: HotkeyProps) {
+  const { MarkersService } = Services;
+  const { hotkey } = props;
+  const description = hotkey.description!;
```

Comment:
we want to avoid bangs where possible, if a description is required in the component i would probably do something like `const description = hotkey.description || ''`